src/nbl/asset/interchange/CPLYMeshWriter.cpp

```diff
@@ -85,37 +85,84 @@ bool CPLYMeshWriter::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
 
     const asset::E_WRITER_FLAGS flags = _override->getAssetWritingFlags(ctx, mesh, 0u);
 
+    auto getConvertedMeshBufferWithNewIndexBuffer = [&]() -> core::smart_refctd_ptr<asset::ICPUMeshBuffer>
+    {
+        auto inputMeshBuffer = *meshbuffers.begin();
+
+        const auto inputIndexType = inputMeshBuffer->getIndexType();
+        const auto inputPrimitiveType = inputMeshBuffer->getPipeline()->getPrimitiveAssemblyParams().primitiveType;
+
+        const void* inputIndices = inputMeshBuffer->getIndices();
+        auto inputIndexCount = inputMeshBuffer->getIndexCount();
+
+        if (inputMeshBuffer->getIndexBufferBinding().buffer && inputPrimitiveType != asset::EPT_TRIANGLE_LIST)
+        {
+            auto copyMeshBuffer = core::smart_refctd_ptr_static_cast<asset::ICPUMeshBuffer>(inputMeshBuffer->clone());
+
+            if (inputPrimitiveType == asset::EPT_TRIANGLE_FAN || inputPrimitiveType == asset::EPT_TRIANGLE_STRIP)
+            {
+                core::smart_refctd_ptr<ICPUBuffer> newIndexBuffer;
+
+                if (inputPrimitiveType == asset::EPT_TRIANGLE_FAN)
+                    newIndexBuffer = IMeshManipulator::idxBufferFromTrianglesFanToTriangles(inputIndices, inputIndexCount, inputIndexType, inputIndexType);
+                else if (inputPrimitiveType == asset::EPT_TRIANGLE_STRIP)
+                    newIndexBuffer = IMeshManipulator::idxBufferFromTriangleStripsToTriangles(inputIndices, inputIndexCount, inputIndexType, inputIndexType);
+
+                asset::SBufferBinding<asset::ICPUBuffer> indexBufferBinding;
+                indexBufferBinding.buffer = newIndexBuffer;
+                indexBufferBinding.offset = 0u;
+
+                copyMeshBuffer->setIndexBufferBinding(std::move(indexBufferBinding));
+                return copyMeshBuffer;
+            }
+        }
+        else 
+            return nullptr;
+    };
+
+    const auto cpuConvertedMeshBuffer = getConvertedMeshBufferWithNewIndexBuffer();
+    const asset::ICPUMeshBuffer* rawConvertedMeshBuffer = cpuConvertedMeshBuffer.get() ? cpuConvertedMeshBuffer.get() : *meshbuffers.begin();
+    const auto primitiveType = rawConvertedMeshBuffer->getPipeline()->getPrimitiveAssemblyParams().primitiveType;
+    const auto indexType = rawConvertedMeshBuffer->getIndexType();
+    uint32_t faceCount = {}; 
+    size_t vertexCount = {};
```

Comment:
all of this can be done with 
```cpp
auto copyMeshBuffer = core::smart_refctd_ptr_static_cast<asset::ICPUMeshBuffer>(inputMeshBuffer->clone());
IMeshManipulator::homogenizePrimitiveTypeAndIndices(&copyMeshBuffer,&copyMeshBuffer+1,asset::EPT_TRIANGLE_LIST,asset::EIT_32BIT);
```

Reply:
corrected in following commit

Reply:
okay, after looking closer at the code, **not entire** code can be replaced with this, but this should be used inside, corrected in following commit

Reply:
why not entire code?

Reply:
because it won't magically create index buffer if there is no such, and it will crash bellow upon 

```C++
if (primitiveType == asset::EPT_POINT_LIST)
        faceCount = 0u;
    else if (indices && indexType != asset::EIT_UNKNOWN)
    {
        header += "element face ";
        header += std::to_string(faceCount) + '\n';
        idxT = indexType;
        const std::string idxTypeStr = idxT == asset::EIT_32BIT ? "uint32" : "uint16";
        header += "property list uchar " + idxTypeStr + " vertex_indices\n";
    }
    else if (primitiveType == asset::EPT_TRIANGLE_LIST)
    {
        forceFaces = true;

        header += "element face ";
        header += std::to_string(faceCount) + '\n';
        idxT = vertexCount <= ((1u<<16) - 1) ? asset::EIT_16BIT : asset::EIT_32BIT;
        const std::string idxTypeStr = idxT == asset::EIT_32BIT ? "uint32" : "uint16";
        header += "property list uchar " + idxTypeStr + " vertex_indices\n";
    }
    else
        faceCount = 0u;
    header += "end_header\n";
```

It will convert index buffer, verticies or indicies, but in that case copying to `indicies`  pointer **each time** is wrong and will be leading to leaks/crashes

Reply:
`homogenizePrimitiveTypeAndIndices` should "magically create an index buffer" if the input meshbuffer's index type is EIT_UNKNOWN and the index buffer binding is nullptr

Reply:
yeah
```C++
const bool iota = indexType==EIT_UNKNOWN||!cpumb->getIndexBufferBinding().buffer;

if (iota)
	correctlyOffsetIndexBufferPtr = iotaUint32Buffer->getPointer();
else
	correctlyOffsetIndexBufferPtr = cpumb->getIndices();

  /// ...
  /// conversion using correctlyOffsetIndexBufferPtr
```

Reply:
so you'll always have an index buffer and it wont crash

are you even using the return value from `homogenize` ?

Reply:
it is a void call

Reply:
I mean, there will be `index` buffer, but the code I pasted above should not check `indicies` pointer after the changes with `homogenize` but index buffer binding as I can see, otherwise the code doesn't make any sense in my opinion

Reply:
well obviously get the "new" index buffer!

Reply:
because it would always launch 

```C++
else if (indices && indexType != asset::EIT_UNKNOWN)
    {
        header += "element face ";
        header += std::to_string(faceCount) + '\n';
        idxT = indexType;
        const std::string idxTypeStr = idxT == asset::EIT_32BIT ? "uint32" : "uint16";
        header += "property list uchar " + idxTypeStr + " vertex_indices\n";
    }
 ```

Reply:
disscused on discord, it's all right